resources/keycloak/custom-user-provider/src/main/java/com/cvmanager/auth/provider/user/CustomUserStorageProvider.java

```diff
@@ -0,0 +1,324 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.stream.Stream;
+
+import org.keycloak.component.ComponentModel;
+import org.keycloak.models.GroupModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.storage.StorageId;
+import org.keycloak.storage.UserStorageProvider;
+import org.keycloak.storage.user.UserLookupProvider;
+import org.keycloak.storage.user.UserQueryProvider;
+import org.keycloak.storage.user.UserRegistrationProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.cvmanager.auth.provider.user.pojos.UserObject;
+
+public class CustomUserStorageProvider implements UserStorageProvider,
+        UserLookupProvider,
+        UserQueryProvider,
+        UserRegistrationProvider {
+
+    private static final Logger log = LoggerFactory.getLogger(CustomUserStorageProvider.class);
+    private KeycloakSession ksession;
+    private ComponentModel model;
+
+    private static String baseUserQuery = """
+                        SELECT
+                keycloak_id,
+                user_id,
+                email,
+                first_name,
+                last_name,
+                created_timestamp,
+                super_user,
+                COALESCE(
+                    jsonb_agg(
+                        jsonb_build_object('org', org_name, 'role', role)
+                    ) FILTER (WHERE org_name IS NOT NULL AND role IS NOT NULL),
+                    '[]'::jsonb
+                ) AS organizations
+            FROM (
+                SELECT
+                    users.keycloak_id,
+                    users.user_id,
+                    users.email,
+                    users.first_name,
+                    users.last_name,
+                    users.created_timestamp,
+                    users.super_user,
+                    org.name AS org_name,
+                    roles.name AS role
+                FROM
+                    public.users
+                LEFT JOIN
+                    public.user_organization AS uo ON uo.user_id = users.user_id
+                LEFT JOIN
+                    public.organizations AS org ON org.organization_id = uo.organization_id
+                LEFT JOIN
+                    public.roles ON roles.role_id = uo.role_id
+            ) AS subquery
+             %s
+            GROUP BY
+                user_id,
+                keycloak_id,
+                email,
+                first_name,
+                last_name,
+                created_timestamp,
+                super_user
+            %s
+            ;
+                        """;
+
+    public CustomUserStorageProvider(KeycloakSession ksession, ComponentModel model) {
+        this.ksession = ksession;
+        this.model = model;
+    }
+
+    private PreparedStatement getBaseUserQuery(Connection c, String where, String end) {
```

Comment:
suggestion(non-blocking): rename this something like `prepareUserQuery` to accurately reflect what the method does.

Reply:
renamed!